keps/sig-scheduling/4832-async-preemption/README.md

```diff
@@ -0,0 +1,417 @@
+
+### Goals
+
+- Improve scheduling throughput when pods require issuing preemptions by making API calls asynchronous
+
+### Non-Goals
+
+- Making the same enhancement for DRA is not a goal of this KEP because it's an under-construction feature yet.
+  - If DRA maintainers want, technically they can along with this KEP. But, at least in this KEP, we don't discuss how.
+
+## Proposal
+
+The preemption plugin makes API calls for the preemption asynchronously after `PostFilter` extension point 
+so that the scheduler can continue to other Pods' scheduling while making API calls for preemption.
+After the preemption goroutine is done, the scheduling for the Pod that triggered the preemption will be retried.
+
+### Risks and Mitigations
+
+#### When kube-apiserver is unstable
+
+When kube-apiserver is unstable and API calls at the preemption goroutine fails frequently,
+the scheduler could make a non-optimal scheduling decision 
+because the scheduler nominates pods at `PostFilter` though, those Pods won't be scheduled on nodes because the preemption API calls fail.
+
+Let's say many mid-priority Pods are making the preemption API calls.
+With the scheduler after this proposal, during the preemption goroutine for them are runnning, 
+the scheduler assumes they'll be scheduled at the Nodes eventually
+that the preemptions are targeting via `.Status.NominatedNodeName`.
+So, other mid-priority or lower priority Pods' scheduling take those preemptor Pods into consideration, 
+which is correct if the preemption goroutine finishes successful actually, while which results in non-best scheduling results otherwise.
+(Higher priority Pods won't be affected; Pods can take place of reserved for lower priority Pods via `.Status.NominatedNodeName`)
+
+But, in the first place though, when kube-apiserver is unstable, the scheduler doesn't behave well 
+because it works with a lot of communication with kube-apiserver.
+Even if the scheduler makes the best scheduling result, the binding API might fail after all.
+
+So, we don't have to pay a special attention to this issue.
+
+## Design Details
+
+To achieve an asynchronous preemption, we will change the preemption plugin's implementation like the following:
+1. The preemption PostFilter plugin calculates the preemption target and nominate the Pod for the Node. (We'll use `AddNominatedPod` API exposed from the scheduling framework to plugins.)
+2. The preemption PostFilter plugin starts the goroutine to make API calls inside, and return success status (= not wait for the goroutine to finish).
+3. The preemption plugin blocks the Pod while the preemption routine is in-progress, using PreEnqueue extension point, so that the target Pod won't be retried during this time.
```

Comment:
Yes, those Pods would be gated, like mentioned in-

> using PreEnqueue extension point

Reply:
If this is another api call then my concern would be we should not introduce more api calls in scheduling cycle and actually our original intension was try to avoid API calls to block the process.

Reply:
Oops. Sorry, my answer is vague.

I said Yes to
> Do you mean gated the Pod?

The answer for ↓ is No.
> Is this another api call?

So, it's just gated at PreEnqueue extension point, not making a new API call.

Reply:
bingo